ouroboros-consensus-protocol/src/ouroboros-consensus-protocol/Ouroboros/Consensus/Protocol/Ledger/HotKey.hs

```diff
@@ -238,9 +230,70 @@ mkEmptyHotKey ::
   => Word64              -- ^ Max KES evolutions
   -> m ()
   -> m (HotKey c m)
-mkEmptyHotKey maxKESEvolutions finalizer = do
+mkEmptyHotKey maxKESEvolutions =
+  mkDynamicHotKey maxKESEvolutions Nothing
+
+mkKESState :: Word64 -> OCert.OCert c -> SL.SignKeyKES c -> Word -> Absolute.KESPeriod -> KESState c
+mkKESState maxKESEvolutions newOCert newKey evolution startPeriod@(Absolute.KESPeriod start) =
+    KESState {
+      kesStateInfo = KESInfo {
+          kesStartPeriod = startPeriod
+        , kesEndPeriod   = Absolute.KESPeriod (start + fromIntegral maxKESEvolutions)
+        , kesEvolution   = evolution
+      }
+      , kesStateKey = KESKey newOCert newKey
+    }
+
+-- | Create a new 'HotKey' that runs a key-producer action on a separate thread.
+-- The key producer action will receive a callback that can be used to pass
+-- keys into the HotKey; the HotKey will dynamically update its internal state
+-- to reflect new keys as they arrive.
+mkDynamicHotKey ::
+     forall m c. (Crypto c, IOLike m)
+  => Word64              -- ^ Max KES evolutions
+  -> Maybe (
+        (OCert.OCert c -> SL.SignKeyKES c -> Word -> Absolute.KESPeriod -> m ())
+        -> m ()
+     )
+  -> m ()
+  -> m (HotKey c m)
+-- | Create a new 'HotKey' that runs a key-producer action on a separate thread.
+-- The key producer action will receive a callback that can be used to pass
+-- keys into the HotKey; the HotKey will dynamically update its internal state
+-- to reflect new keys as they arrive.
+mkDynamicHotKey = mkHotKeyWith Nothing
+
+-- | The most general function for creating a new 'HotKey', accepting an initial
+-- set of credentials, a key producer action, and a custom finalizer.
+mkHotKeyWith ::
+     forall m c. (Crypto c, IOLike m)
+  => Maybe (OCert.OCert c, SL.SignKeyKES c, Word, Absolute.KESPeriod)
+  -> Word64              -- ^ Max KES evolutions
+  -> Maybe (
+        (OCert.OCert c -> SL.SignKeyKES c -> Word -> Absolute.KESPeriod -> m ())
+        -> m ()
+     )
+  -> m ()
+  -> m (HotKey c m)
+mkHotKeyWith initialStateMay maxKESEvolutions keyThreadMay finalizer = do
     varKESState <- newMVar initKESState
 
+    let set newOCert newKey evolution startPeriod =
+          modifyMVar_ varKESState $ \oldState -> do
+            _ <- poisonState oldState
+            return $ mkKESState maxKESEvolutions newOCert newKey evolution startPeriod
+
+    forM_ initialStateMay $ \(newOCert, newKey, evolution, startPeriod) ->
+            set newOCert newKey evolution startPeriod
+
+    finalizer' <- case keyThreadMay of
+      Just keyThread -> do
+        keyThreadAsync <- async
+          (keyThread set)
```

Comment:
Do label this thread. This is mandatory in my opinion, see https://github.com/IntersectMBO/cardano-node/issues/6086